### PR TITLE
Changed output window to be a tab based window.

### PR DESCRIPTION
**Pull request by ChrisMarsh82**
I have added R-Script as it is needed to test these changes however that work is done under  #6953

@dannyparsons Once you have a chance to look through this it would be good to decide how the main output tab should work. It is set up to be a log at the moment i.e. can't change anything. This makes sense in the way the OutputLogger is designed. If this needs to change may need to re-think the structure. Maybe another option would be to decide which tabs are the active recorders? i.e. which tabs will receive the new outputs?

### CODE REVIEW

**Comment by lloyddewit** [comment]
@ChrisMarsh82 As I mentioned yesterday in PR #6953 , I made a new RScript NuGet package that does not include the XUnit dependencies (v. 1.0.4). This should reduce the number of dependencies from 32 to 5.

I suggest that you make a new PR with RScript 1.0.4, this will reduce the number of files in the PR and make it more manageable. You can then close both this PR and PR #6953 .

**Comment by lloyddewit** [comment]
@ChrisMarsh82 I had a quick look through the files. If you make the changes above, then there are about 13 VB files for me to review.
If you can make the new PR this week, then I can aim to finish my review by mid next week. Would that be OK?

**Comment by dannyparsons** [comment]
Looks good. I haven't found any broken or missing functionality. I've tried different kinds of output and didn't find any issues.

@africanmathsinitiative/developers We need more reviews on this as it's a big change. Particularly to check for any breaking changes.

@rdstern What do you think the name of the main tab should be? Maybe `Main`? I don't think `Output` is right.

A few comments:

1. Does there have to be scroll bars for the blocks? It makes it a bit harder to navigate and also reduces the space on the screen. I would be happy just to have the long output without the additional scroll bars.

image

2. After running a command it should scroll to the bottom. It doesn't currently.
3. The sizes seem quite different in this new output than the previous one. The output was a lot bigger than before. Not a problem, but we may need to adjust the default sizes for comments, code and output. 
4. The output window seems to refresh when a dialog is closed without running a command. It should only refresh when a command has been run.
5. I would like to be able to delete an item from the Output tab. For example, the first output I made was way too big and didn't display nicely and make the whole window a bit messy, so I wanted to delete it and do it again.
6. There used to be a right click menu in the Output Window. Some of these options will still be needed.
image
7. `Ctrl + C` when highlighting text doesn't seem to work anymore.
**Other comments that are non essential and could be addressed in future work**
8. When I saved the output as an RFT it was 96MB but it only had one graph in and some text output. That seems quite big. Not sure if that was happening before too. Not a serious problem but the file is bigger than I expected.
9. The multiple tabs works well. It would be good to be able to delete a secondary tab directly. I'm not sure the tab should be deleted when all the items are deleted - can be discussed.
10. It would be nice to be able to scroll up and down the window with the mouse wheel, or keyboard up/down arrows.
11. Select/Deselect all buttons would be nice.

**Comment by N-thony** [comment]
> Looks good. I haven't found any broken or missing functionality. I've tried different kinds of output and didn't find any issues.
> 
> @africanmathsinitiative/developers We need more reviews on this as it's a big change. Particularly to check for any breaking changes.
> 
> @rdstern What do you think the name of the main tab should be? Maybe `Main`? I don't think `Output` is right.
> 
> A few comments:
> 
> 1. Does there have to be scroll bars for the blocks? It makes it a bit harder to navigate and also reduces the space on the screen. I would be happy just to have the long output without the additional scroll bars.
> 
> image
> 
> 2. After running a command it should scroll to the bottom. It doesn't currently.
> 3. The sizes seem quite different in this new output than the previous one. The output was a lot bigger than before. Not a problem, but we may need to adjust the default sizes for comments, code and output.
> 4. The output window seems to refresh when a dialog is closed without running a command. It should only refresh when a command has been run.
> 5. I would like to be able to delete an item from the Output tab. For example, the first output I made was way too big and didn't display nicely and make the whole window a bit messy, so I wanted to delete it and do it again.
> 6. There used to be a right click menu in the Output Window. Some of these options will still be needed.
>    image
> 7. `Ctrl + C` when highlighting text doesn't seem to work anymore.
>    **Other comments that are non essential and could be addressed in future work**
> 8. When I saved the output as an RFT it was 96MB but it only had one graph in and some text output. That seems quite big. Not sure if that was happening before too. Not a serious problem but the file is bigger than I expected.
> 9. The multiple tabs works well. It would be good to be able to delete a secondary tab directly. I'm not sure the tab should be deleted when all the items are deleted - can be discussed.
> 10. It would be nice to be able to scroll up and down the window with the mouse wheel, or keyboard up/down arrows.
> 11. Select/Deselect all buttons would be nice.

12. I have tried to clear the output window from Tools > Clear Output Window and it seems to not work yet
     
image

**Comment by dannyparsons** [comment]
@N-thony have you added comments here? I just see a quote of my comment.

**Comment by N-thony** [comment]
> @N-thony have you added comments here? I just see a quote of my comment.

yes

**Comment by dannyparsons** [comment]
I don't see your comments, can you repost them?

**Comment by N-thony** [comment]
12. I have tried to clear the output window from Tools > Clear Output Window and it seems to not work yet
     
image

**Comment by dannyparsons** [comment]
Thanks, sorry I didn't see that before.

**Comment by lloyddewit** [comment]
> @ChrisMarsh82 As I mentioned yesterday in PR #6953 , I made a new RScript NuGet package that does not include the XUnit dependencies (v. 1.0.4). This should reduce the number of RScript dependencies from 32 to 5.
> 
> I suggest that you make a new PR with RScript 1.0.4, this will reduce the number of files in the PR and make it more manageable. You can then close both this PR and PR #6953 .

@ChrisMarsh82 Please could you close this PR and make a new PR without the XUnit dependencies? If you do this, then I can peer review the code in the new PR, thanks

**Comment by ChrisMarsh82** [comment]
@lloyddewit all dependencies are taken out now

File: instat/Model/Output/clsOutputLogger.vb [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+    ''' <summary>
+    ''' Returns the filtered list given the name. Will create the list if doesnt already exist.
+    ''' </summary>
+    ''' <param name="strListName"></param>
+    ''' <param name="bCreate"></param>
+    ''' <returns></returns>
+    Public Function GetFilteredList(strListName As String, Optional bCreate As Boolean = True) As clsOutputList
+        Dim filteredList As clsOutputList
+        filteredList = _filteredOutputs.Where(Function(x) x.StrName = strListName).FirstOrDefault()
+        If filteredList Is Nothing And bCreate Then
```
Comment (lloyddewit) [inline instat/Model/Output/clsOutputLogger.vb:157]: Will the line above always set `filteredList` to a a found list or a default list? Will `filteredList` ever be `Nothing`?

File: instat/Model/Output/clsOutputLogger.vb [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+    ''' <summary>
+    ''' Checks to see if list name is a valid name for a filtered list
+    ''' </summary>
+    ''' <param name="strListName"></param>
+    ''' <returns></returns>
+    Public Function IsValidFilteredListName(strListName As String) As Boolean
+        If Trim(strListName) Is "" Then
+            Return False
+        End If
+        If _filteredOutputs.Where(Function(x) x.StrName = strListName).Any() Then
```
Comment (lloyddewit) [inline instat/Model/Output/clsOutputLogger.vb:174]:
```suggestion
        If Trim(strListName) Is "" OrElse
            _filteredOutputs.Where(Function(x) x.StrName = strListName).Any() Then
```

File: instat/My Project/app.manifest [hunk 3]
```diff
@@ -0,0 +1,79 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (lloyddewit) [inline instat/My Project/app.manifest:1]: why is this file needed in this PR? why didn't R-Instat require it before?

File: instat/UserControl/ucrOutputPage.vb [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+    Private _bCanRename As Boolean
+    Private _bCanDelete As Boolean
+
+    ''' <summary>
+    ''' Returns all the selected elements
+    ''' </summary>
+    ''' <returns></returns>
+    Public ReadOnly Property SelectedElements() As List(Of clsOutputElement)
+        Get
+            Dim elemets As New List(Of clsOutputElement)
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPage.vb:32]:
```suggestion
            Dim elements As New List(Of clsOutputElement)
```

File: instat/UserControl/ucrOutputPage.vb [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            Case OutputType.TextOutput
+                AddNewTextOutput(outputElement)
+            Case OutputType.ImageOutput
+                AddNewImageOutput(outputElement)
+        End Select
+        pnlMain.VerticalScroll.Value = pnlMain.VerticalScroll.Maximum
+    End Sub
+
+    ''' <summary>
+    ''' Copys selected elenments to clipboard
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPage.vb:128]:
```suggestion
    ''' Copies selected elenments to clipboard
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@ChrisMarsh82 this looks good, I like the way you implemented this. Here are some initial comments.

File: instat/UserControl/ucrOutputPages.vb [hunk 6]
```diff
@@ -0,0 +1,275 @@
...
+        _allOutputPages = New List(Of ucrOutputPage)
+        EnableDisableTopButtons()
+    End Sub
+
+    ''' <summary>
+    ''' Holds options.
+    ''' ToDo InstatOptions should be able to be accessed from anywhere
+    ''' </summary>
+    Public WriteOnly Property clsInstatOptions() As InstatOptions
+        Set(ByVal value As InstatOptions)
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPages.vb:43]: what is the purpose of 'ByVal' here?

File: instat/UserControl/ucrOutputPages.vb [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+        tabControl.TabPages.Remove(tab)
+        tab.Dispose()
+        Dim found As ucrOutputPage = Nothing
+        For Each outputPage In _allOutputPages
+            If outputPage.Tag = tabName Then
+                found = outputPage
+                Exit For
+            End If
+        Next
+        _allOutputPages.Remove(found)
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPages.vb:130]:
```suggestion
        For Each outputPage In _allOutputPages
            If outputPage.Tag = tabName Then
                _allOutputPages.Remove(outputPage)
                Exit For
            End If
        Next
```

File: instat/UserControl/ucrOutputPages.vb [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+            End If
+        Next
+        UpdateTabsInDropDown()
+        EnableDisableTopButtons()
+    End Sub
+
+    Private Sub tbAddToNew_Click(sender As Object, e As EventArgs) Handles tbAddToNew.Click
+        Dim strTabName As String = "New Tab"
+        While Not _outputLogger.IsValidFilteredListName(strTabName)
+            strTabName = strTabName + "1"
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPages.vb:206]:
```suggestion
            strTabName &= "1"
```
This line will make names like "tabName1", "tabName11", "tabName111" etc.
Would "tabName1", "tabName2" etc be better?

File: instat/UserControl/ucrOutputPages.vb [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+
+    Private Sub tbMoveDown_Click(sender As Object, e As EventArgs) Handles tbMoveDown.Click
+        For Each element In _selectedOutputPage.SelectedElements.OrderByDescending((Function(x) x.Id))
+            _outputLogger.GetFilteredList(SelectedTab).MoveElementDown(element)
+        Next
+        RefreshPage()
+    End Sub
+
+    Private Sub tbMoveUp_Click(sender As Object, e As EventArgs) Handles tbMoveUp.Click
+        For Each element In _selectedOutputPage.SelectedElements
```
Comment (lloyddewit) [inline instat/UserControl/ucrOutputPages.vb:237]: Should this explicitly specify order by ascending ID (to be consistent with line 230 above)?

File: instat/Translations.vb [hunk 10]
```diff
@@ -58,7 +58,6 @@ Public Class Translations
 
         'The right mouse button menus for the 6 output windows are not accessible via 
         '    the control lists. Therefore, translate these menus explicitly
-        HandleError(TranslateWinForm.clsTranslateWinForm.TranslateMenuItems(frmMain.ucrOutput.Name, frmMain.ucrOutput.mnuContextRTB.Items, strDbPath, strLanguageCode))
```
Comment (lloyddewit) [inline instat/Translations.vb:61]: Where are the new output windows translated?

File: instat/Translations.vb [hunk 11]
```diff
@@ -127,7 +126,7 @@ Public Class Translations
     ''' <returns>   The full path of the SQLite translations database file. </returns>
     '''--------------------------------------------------------------------------------------------
     Private Shared Function GetDbPath() As String
-        Dim strTranslationsPath As String = String.Concat(System.Windows.Forms.Application.StartupPath, "\translations")
+        Dim strTranslationsPath As String = String.Concat(System.Windows.Forms.Application.StartupPath, "/translations")
```
Comment (lloyddewit) [inline instat/Translations.vb:129]: I'm assuming that `/` works with both Windows and Linux, is that correct?

**Review by lloyddewit — CHANGES_REQUESTED** [review]
I finished my review

File: instat/Model/Output/clsOutputList.vb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            _output = value
+        End Set
+    End Property
+
+    ''' <summary>
+    ''' Changes position of element - Down
+    ''' </summary>
+    ''' <param name="outputElement"></param>
+    Public Sub MoveElementDown(outputElement As clsOutputElement)
+        If outputElement.Id < _output.Count Then
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Model/Output/clsOutputList.vb:64]: They do start at 1
outputElement.Id = filteredList.Output.Count + 1

File: instat/Model/Output/clsOutputList.vb [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            outputElement.Id += 1
+        End If
+    End Sub
+
+    ''' <summary>
+    ''' Changes position of element - Up
+    ''' </summary>
+    ''' <param name="outputElement"></param>
+    Public Sub MoveElementUp(outputElement As clsOutputElement)
+        If outputElement.Id > 1 Then
```
[the remark this answers is not in the document]
Comment (ChrisMarsh82) [reply instat/Model/Output/clsOutputList.vb:75]: They do start at 1
outputElement.Id = filteredList.Output.Count + 1

File: instat/Model/Output/clsOutputLogger.vb (hunk 1 above)
Comment (ChrisMarsh82) [reply instat/Model/Output/clsOutputLogger.vb:157]: It calls GetFilteredList if it exists or not. The optional bCreate decides if it should be created or return nothing if it doesn't exist

File: instat/UserControl/ucrOutputPage.vb (hunk 4 above)
Comment (ChrisMarsh82) [reply instat/UserControl/ucrOutputPage.vb:32]: Changed

File: instat/UserControl/ucrOutputPages.vb (hunk 6 above)
Comment (ChrisMarsh82) [reply instat/UserControl/ucrOutputPages.vb:43]: This is added automatically by Visual Studio. Then visual studio tells you it's not needed. Will take it out.

File: instat/UserControl/ucrOutputPages.vb (hunk 9 above)
Comment (ChrisMarsh82) [reply instat/UserControl/ucrOutputPages.vb:237]: Added in

File: instat/My Project/app.manifest (hunk 3 above)
Comment (ChrisMarsh82) [reply instat/My Project/app.manifest]: removed

Comment (ChrisMarsh82) [reply instat/My Project/app.manifest]: removed

File: instat/Translations.vb (hunk 11 above)
Comment (ChrisMarsh82) [reply instat/Translations.vb:129]: yep

File: instat/Translations.vb (hunk 10 above)
Comment (ChrisMarsh82) [reply instat/Translations.vb:61]: Added translations back in

**Comment by ChrisMarsh82** [comment]
Can't resolve the merge issue. Tried to use command line but it either wanted me to create a new pull request with all the changes duplicated or just didn't notice the conflict. - frmOutputWindow should just be deleted.

@dannyparsons  I fixed some of the issues however the refreshing when the dialog closes, scroll bars on the elements and font sizes don't seem to be an issue for me. I changed the dialogs running on 100% zoom to see if this makes any difference. Hopefully we can chat first thing tomorrow about it.

**Comment by dannyparsons** [comment]
Could you maybe manually undo the changes to the conflicting files? It's just two and the changes are small. Then we could merge and redo the changes as needed.

**Pull request merged**